### ISSUE TITLE
JsxLexers doesn't support destructing objects

### ISSUE THREAD

**Issue by jare-talend** [i18next/i18next-parser#96]
- Is this a bug or a feature request? bug
- What version are you running (`0.x` or `1.x`)? 1.x
- Which runner (cli, gulp...)? cli

## What I'm trying to do?
I want to parse a file with destructing objects in.

e.g.

src/test.js

```
import React from 'react';

export default function test({ foo, ...rest }) {
	return <div {...rest} />;
}
```

## How am I doing it?

*Replace this section with your config file or cli command*

Here is my config file:

```
{
  output: "locales",
  "lexers": {
    "js": ["JsxLexer"]
  }
}
```

The command I'm running:

```
i18next src/**/*.js -c ./config.json
```

## What was the result?

i18next triggers an error because acorn-jsx doesn't support destructing objects
https://github.com/RReverser/acorn-jsx/issues/75 
```
$ i18next ./src/**/*.js -c ../../config.json

  i18next Parser
  --------------
  Input:  ./src/**/*.js
  Output: locales

  [read]  test.js
{}
node_modules/acorn/dist/acorn.js:2750
  throw err
  ^

SyntaxError: Unexpected token (3:36)
    at Parser.pp$4.raise (node_modules/acorn/dist/acorn.js:2748:13)
    at Parser.pp.unexpected (node_modules/acorn/dist/acorn.js:644:8)
    at Parser.pp$3.parseIdent (node_modules/acorn/dist/acorn.js:2703:10)
    at Parser.pp$3.parsePropertyName (node_modules/acorn/dist/acorn.js:2487:105)
    at Parser.pp$3.parseProperty (node_modules/acorn/dist/acorn.js:2417:8)
    at Parser.pp$3.parseObj (node_modules/acorn/dist/acorn.js:2371:23)
    at Parser.pp$2.parseBindingAtom (node_modules/acorn/dist/acorn.js:1674:19)
    at Parser.pp$2.parseMaybeDefault (node_modules/acorn/dist/acorn.js:1714:23)
    at Parser.pp$2.parseBindingList (node_modules/acorn/dist/acorn.js:1699:25)
    at Parser.pp$1.parseFunctionParams (node_modules/acorn/dist/acorn.js:1223:22)
error Command failed with exit code 1
```
*Replace this with a description of the result you are getting*

## What was I expecting?

parse the file without an error

**Comment by karellm** [comment]
@MynockSpit I really appreciate your help. Thanks a lot, it's so much more fun to have contributors than doing it all alone 😄 

While you are investigating the other issue, if this seems like an easy fix that would be great if you could tackle this too. Otherwise I will have a look later this week.

**Comment by MynockSpit** [comment]
This could be a tough one. Means either a) we need to bug the authors of (or make a PR for) acorn-jsx b) switch parsers or c) write a pre-processor that removes the ...'s from files (and adds them to strings if the appear in keys?

Alternatively, we just say we don't support `...` like it seems like React did. 😕

**Comment by jare-talend** [comment]
@MynockSpit 

A solution possible: inject the plugins acorn-object-rest-spread in acorn. It seems manage destructing objects. I didn't test if they are a impact with the JSX parser.

```
const acorn = require('acorn');
require('acorn-jsx/inject')(acorn);
require('acorn-object-rest-spread/inject')(acorn);

acorn.parse(`export default function test({ foo, ...rest }) {
	return <div {...rest} />;
}`, {
sourceType: 'module',
  plugins: { jsx: true, objectRestSpread: true}
});
```

**Comment by karellm** [comment]
@jare-talend or @MynockSpit Can you make a PR?

**Comment by MynockSpit** [comment]
I'll try to take a look at this this weekend.

**Comment by ledfusion** [comment]
It looks like decorators are not supported as well (easily workaround-able), but static class properties also fail to parse.

```js
export default class ClubCalendar extends Component {

	static propTypes = {
		navigation: PropTypes.object.isRequired,
		t: PropTypes.func.isRequired,
		app: PropTypes.object.isRequired,
		dispatch: PropTypes.func.isRequired
	};

	static navigationOptions = ({ navigation, screenProps }) => ({
		title: screenProps.t('Calendari')
	});

         // ...
}
```

```
SyntaxError: Unexpected token (34:18)
    at Parser.pp$4.raise (/Users/jordi/git/app/node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:2756:13)
```

Is there an easy way to add a flag for them? 
Thank you!

**Comment by karellm** [comment]
@MynockSpit Is there any update on your end? If you are too busy I might take a stab at it this week.

**Comment by ledfusion** [comment]
I had some advance by using this fragment in the config file:

```javascript
module.exports = {
        ...
	lexers: {
		....

		js: [
			{
				lexer: 'JavascriptLexer',
				acorn: {
					plugins: {
						stage3: true,
						jsx: true,
						// objectRestSpread: true,
						staticClassPropertyInitializer: true,
						es7: true
					}
				}
			}
		]
	}
};

```

This config is tightly coupled with acorn, so I think that there should be at least a clear example in the Readme

**Comment by ledfusion** [comment]
However, the problem is still not solved.

Now, the parser complains about the plugins not being installed. 

So I installed them manually:

```sh
yarn add -D acorn-stage3 acorn-static-class-property-initializer acorn-es7 acorn-object-rest-spread
```

But still, the same.

So digging into the code, I saw that `i18next-parser` would need to register them before they can be used. I can make a pull request with the equivalent of:

javascript-lexer.js

```javascript
  extract(content) {
    const that = this

      var injectAcornObjectRestSpread = require('acorn-object-rest-spread/inject');
      var injectAcornStaticClassPropertyInitializer = require('acorn-static-class-property-initializer/inject');
      
      require('acorn-es7')(acorn);
      injectAcornStaticClassPropertyInitializer(acorn);
      acorn = require('acorn-stage3/inject')(acorn);
      injectAcornObjectRestSpread(acorn);

      walk.simple(
      // ...
```

But then, acorn fails like so:

```sh
TypeError: baseVisitor[type] is not a function
    at c (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:29:22)
    at Object.base.ClassBody (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:421:5)
    at c (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:29:22)
    at Object.base.Class (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:414:3)
    at c (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:29:22)
    at Object.base.ClassDeclaration.base.ClassExpression (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:410:80)
    at c (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:29:22)
    at Object.skipThrough (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:183:37)
    at c (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:29:22)
    at Object.base.Program.base.BlockStatement (/Users/jordi/git/federapps/llicencia-equip/node_modules/acorn/dist/walk.js:195:5)
```

Acorn doesn't know about "ClassProperty" and the process fails. 

It seems that there are many pieces to work out yet :(

**Comment by ledfusion** [comment]
PR #108 is ready

**Comment by karellm** [comment]
This is fixed in `1.0.0-beta22`. Thanks for all the contributions!

**Comment by JounQin** [comment]
It seems `2.1.0` breaking usage of `<div {...props}></div>`?
@karellm @mysticatea @marijnh @RReverser 
Downgrading to `2.0.1` works to me.
